3d material datadefined

vcloarec opened the pull request:
Phong and Gooch materials can be data defined

![hotSpot](https://user-images.githubusercontent.com/7416892/95643466-5a78fb80-0a7d-11eb-9b64-0858b9614034.jpeg)
![categorized](https://user-images.githubusercontent.com/7416892/95643469-5cdb5580-0a7d-11eb-858f-77c62502a71e.jpeg)

nyalldawson committed:
Add interface API for geocoders

rldhont committed:
Processing - GDAL - Raster clip by extent - override CRS

willcohen committed:
[FEATURE] Improved color vision deficiency sim

This modifies the previous support for grayscale
and LMS-based simulation for protanopia and
deuteranopia, and brings it in line with the
methodology currently used in Chromium and Firefox
(https://bugs.chromium.org/p/chromium/issues/detail?id=1003700,
https://bugzilla.mozilla.org/show_bug.cgi?id=1655053).

QGIS now uses updated grayscale luminance
calculations (renamed to achromatopsia), a
precomputed protanopia matrix (renamed from
protanope), a precomputed deuteranopia matrix
(renamed from deuteranope), and an additional mode
for tritanopia using a similarly precomputed matrix.

This commit addresses issue #29760.

willcohen committed:
Revert CVD function names to avoid API breakage

willcohen committed:
PreviewGrayScale->PreviewModeGrayscale, captions

willcohen committed:
Revert ordering of PreviewMode's Grayscale/Mono

willcohen committed:
Align tritanopia tooltip with text string

DelazJ committed:
Update more geometry overlay functions help

DelazJ committed:
Update overlay_nearest function help and examples

DelazJ committed:
Fix wrong copy-paste

elpaso committed:
Landing page: respect drawing order from TOC

Fix WMS drawing order and icon alin for
rule groups.

vcloarec committed:
fix map setting temporal be active when updating time controller

nyalldawson committed:
Remove non-functional loadDefaultStyle from QgsPointCloudLayer

We should use the base class version for point cloud layers

nyalldawson committed:
Don't show unwanted error messages relating to loading default metadata/style for pointcloud layers

nyalldawson committed:
Avoid trying to load unnecessary styling for temporary snap layer

nyalldawson committed:
Another default style handling fix for point cloud layers

nyalldawson committed:
Add standard actions to save/restore default metadata for point cloud layers

willcohen committed:
Revert grayscale luminance constants

While dcc5a1f9 updated the CVD grayscale simulation
to match Chromium and Firefox, Nguyen (2017)
(https://doi.org/10.1109/CVPR.2017.627) writes that
those constants incorrectly assume a conversion from
the NTSC color space. QGIS was previously using
constants from sRGB, and this commit reverts them,
adding decimal points to correspond to Nguyen.

nyalldawson committed:
Add a hacky configuration page for point cloud 2d renderer

willcohen committed:
Modify tritanopia 'since' to reflect 3.18

nyalldawson committed:
Hacky save/restore 2d renderer settings for point clouds

willcohen committed:
Remove references to fax and photocopy

nyalldawson committed:
Fix style save options in layer properties dialog

sklencar committed:
[QgsQuick] Added soft and hard constraints in the form (#38742)

* [QgsQuick] Added soft and hard constraints in the form

* [QgsQuick] Fixed field description label in the QgsQuick form

* [QgsQuick] code layout

Co-authored-by: vsklencar <viktor.sklencar@lutraconsulting.co.uk>

elpaso committed:
Merge pull request #39625 from elpaso/server-landingpage-fix-wms-layer-order

Landing page: respect drawing order from TOC

elpaso committed:
Fix a couple of QT 5.15 deprecation warnings

alexbruy committed:
[processing] correctly escape parentheses in GDAL command (fix #39525)

troopa81 committed:
[processing] Add missing parameter when reading QgsProcessingParameterField from string

wonder-sk commented on src/3d/materials/qgsgoochmaterialsettings.cpp:
```diff
@@ -113,3 +127,143 @@ Qt3DRender::QMaterial *QgsGoochMaterialSettings::toMaterial( QgsMaterialSettings
+
+
+  QByteArray array;
+  array.resize( sizeof( float ) * 12 );
```
do we need to use 4-byte float for each color component? why not just 1-byte `unsigned char` for each color component? The colors will be translated to 1-byte values in the final R,G,B color anyway...

wonder-sk commented on src/3d/materials/qgsgoochmaterialsettings.cpp:
```diff
@@ -113,3 +127,143 @@ Qt3DRender::QMaterial *QgsGoochMaterialSettings::toMaterial( QgsMaterialSettings
+
+int QgsGoochMaterialSettings::dataDefinedByteStride() const
+{
+  return 9 * sizeof( float );
```
why the byte stride is 9 * float? Why not 12 * float?

wonder-sk commented on src/3d/materials/qgsphongmaterialsettings.cpp:
```diff
@@ -131,3 +142,117 @@ void QgsPhongMaterialSettings::addParametersToEffect( Qt3DRender::QEffect *effec
+  diffuseAttribute->setCount( vertexCount );
+  geometry->addAttribute( diffuseAttribute );
+
+  Qt3DRender::QAttribute *ambiantAttribute = new Qt3DRender::QAttribute( geometry );
```
typo here and in few other places: ambiant -> ambient

wonder-sk commented on src/app/3d/qgsphongmaterialwidget.cpp:
```diff
@@ -81,5 +102,13 @@ QgsAbstractMaterialSettings *QgsPhongMaterialWidget::settings()
   m->setSpecular( btnSpecular->color() );
   m->setShininess( spinShininess->value() );
+
+  QgsPropertyCollection dataDefinedPropertie;
```
typo: Propertie -> Properties

nyalldawson committed:
Fix logic in QgsUserInputWidget which incorrectly deletes other
widgets whenever one existing widget is deleted

Fixes calls to iface.addUserInputWidget() does not show any widget if
an existing user input widget is already shown

nyalldawson committed:
Make tracer less debug noisy

nyalldawson committed:
Fix reshape line tool fails when reshape line partially exactly
overlaps geometry

nyalldawson committed:
Reshape tool should only operate on the selected features, if a selection is present

Fixes #23088

elpaso committed:
Server add min/max scale to symbol json export

elpaso committed:
Test for server add min/max scale to symbol json export

vcloarec committed:
3D material color data defined

typo


typo


replace float by byte

elpaso committed:
Fix wms getfeatureinfo json with composite pks

Fixes #39682

alexbruy committed:
Merge pull request #38841 from rldhont/processing-gdal-clipbyextent

[Processing][GDAL] Clip Raster by extent - override CRS

alexbruy committed:
[processing] test for newly added parameter in Clip Raster by Extent
(followup 41dc20c2c1)

vcloarec committed:
fix unused


fix unused

elpaso committed:
Different code path for unsupported providers

alexbruy committed:
Merge pull request #39687 from alexbruy/gdal-clip-test

[processing] test for newly added parameter in Clip Raster by Extent (followup 41dc20c2c1)

rduivenvoorde committed:
Handle 'inbuilt:' vector uri in iface.addVectorLayer() method

nyalldawson committed:
Spelling, since updates

nyalldawson committed:
Spelling

elpaso committed:
Merge pull request #39673 from elpaso/server-wms-getlegendgraphics-json-symbolscale

Server wms getlegendgraphics json symbolscale

elpaso committed:
Merge pull request #39684 from elpaso/bugfix-gh39682-server-getfeatureinfo-json-composite-pks

Fix wms getfeatureinfo json with composite pks

vcloarec committed:
add tests and typo

vcloarec committed:
fix data buffer

tschmetzer committed:
Rename variables for a more intuitive documentation

vcloarec committed:
test for 3d data defined rendering

vcloarec committed:
add image control for travis

vcloarec committed:
color depending on feature on test

vcloarec committed:
travis control image
